Sources/MongoSwift/BSON/Document+Sequence.swift

```diff
@@ -0,0 +1,121 @@
+import Foundation
+import libmongoc
+
+/// An extension of `Document` to make it conform to the `Sequence` protocol.
+/// This allows you to iterate through the (key, value) pairs, for example:
+/// ```
+/// let doc: Document = ["a": 1, "b": 2]
+/// for (key, value) in doc {
+///     ...
+/// }
+/// ```
+extension Document: Sequence {
+    /// Returns a `DocumentIterator` over the values in this `Document`. 
+    public func makeIterator() -> DocumentIterator {
+        guard let iter = DocumentIterator(forDocument: self) else {
+            preconditionFailure("Failed to initialize an iterator over document \(self)")
+        }
+        return iter
+    }
+}
+
+/// An iterator over the values in a `Document`. 
+public class DocumentIterator: IteratorProtocol {
+    /// the libbson iterator. it must be a `var` because we use it as
+    /// an inout argument
+    internal var iter: bson_iter_t
+
+    /// Initializes a new iterator over the contents of `doc`. `doc` must remain alive for 
+    /// the lifetime of the iterator. Returns `nil` if the key is not found, or if an iterator 
+    /// cannot be created over `doc` due to an error from e.g. corrupt data.
```

Comment:
What happens if `doc` does not remain alive for the lifetime of the iterator?

I imagine there's a risk we'd have dangling pointers and could crash trying to iterate further. Would it make sense to actually save a reference to it here and prevent it from being freed due to reference counting?

Similarly, that might avoid in-place updates where Document decides to duplicate its internal storage if there are multiple references to it.